### DISCUSSION TITLE
Snippet loader for lua

### DISCUSSION THREAD

**Discussion by L3MON4D3** [L3MON4D3/LuaSnip#341] [Ideas]
Considering the functionality in [this wiki-entry](https://github.com/L3MON4D3/LuaSnip/wiki/Nice-Configs#split-up-snippets-by-filetype-load-on-demand-and-reload-after-change-first-iteration) ~will soon be deprecated~ is now deprecated, I'd like to add a loader for lua-snippets which is capable of loading them on-demand (like `lazy_load` from the current loaders) 

I think a format very close to that of snipmate, eg. a directory containing `ft.lua`-files, each returns a table, containing keys `snippets`, `autosnippets`, would be pretty good, as it's almost compatible with the current solution.
Maybe we could also autodetect these collections by requiring that for autodetection they be named eg. `luasnippets` and be in a directory in `path` (just `lua/snippets`+`require` would work  only as long as just one luasnippet-collection is present).

Reload after changes should be implementable by invalidating previously loaded snippets (I think we'll need a way to clean up invalidated snippets :D) and re-adding the changed file.

What do you guys think?

**Comment by leiserfg** [comment]
Is not that the same as putting the scripts to load the snippets on `ftplugin`?

**Reply by L3MON4D3** [reply]
True, that would also work, but then everybody'd need to do that for each filetype they have snippets for.
I think it'd be nice to offer that functionality+more (reload, editing current filetype) ootb

**Reply by leiserfg** [reply]
With reloading you mean reload only a `ft` or all the snippets? 1st requires improving the invalidation (probably by actually removing the snippets from the snippets registry)  second just requires a function to nuke the whole registry. 
For the 1st I think we will need a way to uniquely address a snippet in the registry so they can be  "surgically " removed.

**Reply by L3MON4D3** [reply]
Just the snippets loaded from the edited file.
I was thinking of keeping references to the snippets loaded from each file, if the file should be reloaded we can invalidate all of those and just add the new ones.
Youre right, the invalidation needs to be improved, we dont want to iterate over 100s of useless snippets each expand (autoexpand😬😬).
We could count the number of invalidated snippets, once it reaches a threshold actually remove the invalidated ones.

**Reply by kunzaatko** [reply]
I do not know what the actual implementation of this would look like, but I think it would be more flexible to expose some kind of `register_snippets` function, that can be called more than once and that does not directly affect the snippets table. Then the `lazy_load` capability, can be handled by pointing the secondary snippets table (affected by `register_snippets`) to the actual primary snippets table, which would do all the setting of snippets meta-tables only when something is added.
If I am totally off with my thought, then let me know :)

**Reply by kunzaatko** [reply]
Then sourcing the file with the `register_snippets` call would only affect the snippets that were changed in the file. Right?

**Reply by leiserfg** [reply]
The "new way" of storing the snippets @L3MON4D3  is working on is more like a register with several views, I'm not sure if using tables with `__index` magic will be nice, as it will require a lot of logic around. The faux deletion plus eventual cleanup seems more appropriate.  Given that @L3MON4D3  is using `setfenv` already he could just track which snippets belong to each file and clean them all on reload.

**Reply by L3MON4D3** [reply]
Yes, It's elegant for the current internal layout, but using metatables and the like would require some working around for the one we'll be using soon.

But the idea of reloading for regular, non-lazy_loaded snippets is great, we could modify `add_snippets` to accept a key, calling it again with the same key invalidates previously loaded snippets and adds the new ones.
Should also not be too much to implement, we basically have everything necessary in `loaders/from_lua`, just have to move it to `add_snippets`/`session`.

**Reply by galou** [reply]
@L3MON4D3 the wiki-entry is already deprecated. Are you talking about [this](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#lua-snippets-loader)? I tried to configure my snippets as written there but didn't succeed. Should I wait a bit for this to be implemented?

My current attempt is `lua/plugin_setup/luasnip.lua` and `lua/luasnippets/*.lua` in https://github.com/galou/dotvim, commit [2252142](https://github.com/galou/dotvim/commit/2252142d4b6620291181ea61ac55a49abf9be7a0)](https://github.com/galou/dotvim/commit/2252142d4b6620291181ea61ac55a49abf9be7a0). I use LuaSnip commit 80e68242.

**Reply by L3MON4D3** [reply]
Ah, right, I'll edit the comment, maybe I can point to an older version in the wiki.

The loader is implemented in your commit, you'll just have to move `lua/luasnippets` to `luasnippets`. I'll add an example for that path in the doc, reading it again it isn't that clear.

**Reply by galou** [reply]
I understand my error now. I interpreted "runtime path" as lua runtime path not vim runtime path.
